clear-site-data/support/test_utils.sub.js

```diff
@@ -245,6 +245,60 @@ var TestUtils = (function() {
    */
   TestUtils.populateStorage = populate.bind(this, TestUtils.STORAGE);
 
+  /**
+   * Wait for a postMessage from an iFrame, specifically a message that communicates
+   * the result of a request that is served by the service worker
+   * @private
+   */
+  TestUtils.getServiceWorkerControlledResponseBody = function() {
+    return new Promise(function(resolve) {
+      const cb = (m) => {
+        if (m.data && m.data.subject === "service-worker-page-request") {
+          window.removeEventListener("message", cb);
+          resolve(m.data.body);
+        }
+      };
+      window.addEventListener("message", cb);
+    });
+  }
+
+  /**
+   * Wait for a postMessage from an iFrame, specifically a message that communicates
+   * that a controllerchange event has happened.
+   * @private
+   */
+  TestUtils.waitForControllerChangeEvent = function() {
+    return new Promise(function(resolve) {
+      const cb = (m) => {
+        if (m.data && m.data.subject === "service-worker-page-controllerchange") {
+          window.removeEventListener("message", cb);
+          resolve(m.data.hasController);
+        }
+      };
+      window.addEventListener("message", cb);
+    });
+  }
+
+  /**
+   * Returns a promise that resolves when the registration has an active worker
+   * @param reg The service worker registration object
+   * @return a promise that resolves when the worker becomes active
+   * @private
+   */
+  TestUtils.waitForServiceWorkerActive = function(reg) {
+    return new Promise((resolve) => {
+      (function pollForControllingWorker() {
+        if (reg.active && reg.active.state === 'activated') {
+          resolve();
+        } else {
+          step_timeout(() => {
+            pollForControllingWorker();
+          }, 100);
```

Comment:
Why do we need this polling mechanism?  Most tests uses the statechange event on the ServiceWorker object.  There is an existing helper method that uses it call wait_for_state() or something.

Reply:
@wanderview 

> Why do we need this polling mechanism?

No good reason, but my original rationale was:
1. My custom polling looks for a controller property on the registration, rather than waiting for the state of a specific worker. By the time this callback runs, we can’t assume that the worker is still installing, so we have to try and pick a worker, which feels a bit uglier than just polling for registration state, but this is pretty subjective.
2. Calling `wait_for_state()` also felt ugly because it requires passing in a test object. As far as I know, the only way to get this test object is to call `promise_test.` We don't actually want to make any assertions at this point, we are only calling `promise_test` to get the test object so we can call `wait_for_state()`. Again, purely subjective.

I updated the PR to use `wait_for_state()`, which I think looks better now, so thanks